File: frontend/lib/api/review.ts
import { apiClient } from '@/lib/api/client'

export type FlaggedMessage = {
  message_id: string
  conversation_id: string
  bot_response: string
  user_query: string
  intent?: string | null
  intent_confidence?: number | null
  feedback_score?: number | null
  created_at: string
  retrieved_documents?: Array<Record<string, unknown>> | null
  reason: string
}

export type FlaggedMessageListResponse = {
  items: FlaggedMessage[]
  total: number
  skip: number
  limit: number
}

export type ReviewStatistics = {
  total_flagged: number
  pending: number
  approved: number
  rejected: number
  added_to_kb: number
  by_intent: Record<string, number>
  average_review_time_minutes?: number | null
  top_reviewers: Array<{ reviewer_id: string; reviewed_count: number }>
}

export type ReviewActionRequest = {
  corrected_content?: string
}

export type AddToKnowledgeBaseRequest = {
  category: string
  title: string
}

export async function getFlaggedMessages(params: {
  skip?: number
  limit?: number
  intent?: string
}): Promise<FlaggedMessageListResponse> {
  const response = await apiClient.get<FlaggedMessageListResponse>('/api/v1/admin/review/flagged', {
    params,
  })
  return response.data
}

export async function approveMessage(
  messageId: string,
  data: ReviewActionRequest
): Promise<boolean> {
  const response = await apiClient.post<boolean>(`/api/v1/admin/review/${messageId}/approve`, data)
  return response.data
}

export async function rejectMessage(messageId: string): Promise<boolean> {
  const response = await apiClient.post<boolean>(`/api/v1/admin/review/${messageId}/reject`)
  return response.data
}

export async function addToKnowledgeBase(
  messageId: string,
  data: AddToKnowledgeBaseRequest
): Promise<string> {
  const response = await apiClient.post<string>(`/api/v1/admin/review/${messageId}/add-to-kb`, data)
  return response.data
}

export async function getReviewStatistics(
  params: {
    date_from?: string
    date_to?: string
  } = {}
): Promise<ReviewStatistics> {
  const response = await apiClient.get<ReviewStatistics>('/api/v1/admin/review/statistics', {
    params,
  })
  return response.data
}

File: backend/app/schemas/review.py
"""Schemas for continuous-learning review workflows."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.knowledge_base import KnowledgeCategory


class FlaggedMessageResponse(BaseModel):
    """Flagged assistant message with reviewer context."""

    message_id: UUID
    conversation_id: UUID
    bot_response: str
    user_query: str
    intent: str | None = None
    intent_confidence: float | None = Field(default=None, ge=0, le=1)
    feedback_score: int | None = Field(default=None, ge=-1, le=1)
    created_at: datetime
    retrieved_documents: list[dict[str, Any]] | None = None
    reason: str


class FlaggedMessageListResponse(BaseModel):
    """Paginated flagged message list."""

    items: list[FlaggedMessageResponse]
    total: int
    skip: int
    limit: int


class ReviewActionRequest(BaseModel):
    """Approve/reject payload with an optional correction."""

    corrected_content: str | None = Field(default=None, min_length=1)


class AddToKnowledgeBaseRequest(BaseModel):
    """Request to convert an approved chat answer to a KB document."""

    category: KnowledgeCategory
    title: str = Field(min_length=1, max_length=255)


class ReviewerStatistic(BaseModel):
    """Review count by reviewer."""

    reviewer_id: UUID
    reviewed_count: int


class ReviewStatistics(BaseModel):
    """Aggregated review queue and completion metrics."""

    total_flagged: int
    pending: int
    approved: int
    rejected: int
    added_to_kb: int
    by_intent: dict[str, int]
    average_review_time_minutes: float | None = None
    top_reviewers: list[ReviewerStatistic] = Field(default_factory=list)
